compiler/rustc_typeck/src/check/demand.rs

```diff
@@ -616,10 +616,28 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             }
             _ if sp == expr.span && !is_macro => {
                 if let Some(steps) = self.deref_steps(checked_ty, expected) {
+                    let expr = expr.peel_blocks();
+
                     if steps == 1 {
-                        // For a suggestion to make sense, the type would need to be `Copy`.
-                        if self.infcx.type_is_copy_modulo_regions(self.param_env, expected, sp) {
-                            if let Ok(code) = sm.span_to_snippet(sp) {
+                        if let hir::ExprKind::AddrOf(_, _, expr) = expr.kind {
+                            // If the expression has `&`, removing it would fix the error
+                            if let Ok(code) = sm.span_to_snippet(expr.span) {
+                                let message = "consider removing the `&`";
+                                let suggestion = format!("{}", code);
+                                return Some((
+                                    expr.span,
```

Comment:
This `expr.span` points at the span of everything but the `&` because you're shadowing `expr`. Here you want the _outer_ `expr.span` that _includes_ the `&`. And because of how you're checking this, we can get rid of the `span_to_snippet` and instead make a span covering just the `&` and with a suggestion of `String::new()`.

Reply:
> And because of how you're checking this, we can get rid of the span_to_snippet and instead make a span covering just the &

Hm, so just the span of `&` can be obtained with `shrink_to_lo`, but that would be incorrect when the `AddrOf` is mutable and we have `&mut`. Is there a way to get the span of `&mut` from an `AddrOf` expr?

> with a suggestion of String::new().

Can I make this a `run-rustfix` test with an empty suggestion?

Reply:
Nvm, figured it out